### PR TITLE
Feature - Add extension to support max apk size

### LINKED ISSUE #48: [Feature Request] Fail build on "maxApkSize"
## Question:
Can we add an integer such as "maxApkSize" in an extension in order to fail builds that go over a certain size?

## Example:
```
apkSize {
  maxApkSize = 5800000
}
```

This will help developers that are developing on their local machines not only make better decisions on the code that they write but the libraries they use.

### PR DESCRIPTION
**Pull request by jaredsburrows**
@vanniktech closes #48 

Add extension to support max apk size

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,8 +65,22 @@ Total time: 33.017 secs
...
 Those CSV files can for instance be used to configure the [Jenkins Plot Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Plot+Plugin), to see the apk size for each build in a graph.
 
+## Configuration
+
+Apksize is configurable via a Gradle extension (shown with default values):
+
+in `app/build.gradle`:
+```groovy
+dexcount {
+    maxApkSize = 5800000
```
Comment (jaredsburrows) [inline README.md:75]: Added appropriate documentation.

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy [hunk 2]
```diff
@@ -33,9 +35,20 @@ class ApkSizeTask extends DefaultTask {
                 appendableStream.println(bytes + "," + kiloBytes + "," + megaBytes)
             }
         }
+
+        failBuildMaxMethods(apkSize, extension.maxApkSize)
```
Comment (jaredsburrows) [inline src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy:39]: Run this method last, so the user(developer) can see the exception after the reports have printed to STDOUT and saved to disk.

**Review by vanniktech — COMMENTED** [review]
Any reason for changing all the quotes as well as the parenthesis?

**Comment by jaredsburrows** [comment]
It was mixed in some areas, I was just trying to make it consistent.

File: src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy [hunk 3]
```diff
@@ -6,21 +6,22 @@ import spock.lang.Specification
...
         given:
-        ApkSizeTask task = project.task 'apkSize', type: ApkSizeTask
-        def parent = new File('./src/test/assets')
-        task.apk = new File(parent, 'test.xml')
-        task.outputFile = new File(parent, 'test-result.csv')
+        ApkSizeTask task = project.task "apkSize", type: ApkSizeTask
+        def parent = new File("./src/test/assets")
+        task.apk = new File(parent, "test.xml")
+        task.outputFile = new File(parent, "test-result.csv")
+        task.extension = new ApkSizeExtension(maxApkSize: 500)
```
Comment (jaredsburrows) [inline src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy:24]: Add fake size for test.

**Comment by jaredsburrows** [comment]
@vanniktech Let me know what you think.

File: README.md [hunk 4]
```diff
@@ -65,8 +65,22 @@ Total time: 33.017 secs
...
+}
+```
+
+Configurable flags:
+- `maxApkSize`: Set the max APK size. Greater than specified size will fail the build. Default is -1.
```
Comment (vanniktech) [inline README.md]: should say in which unit the size is

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeExtension.groovy [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.vanniktech.android.apk.size
+
+class ApkSizeExtension {
+  /**
+   * Set the max APK size. Greater than specified size will fail the build.
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/apk/size/ApkSizeExtension.groovy]: should say in which unit the size is

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy [hunk 6]
```diff
@@ -3,27 +3,29 @@ package com.vanniktech.android.apk.size
...
 import org.gradle.api.DefaultTask
+import org.gradle.api.GradleException
 import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.TaskAction
 
 class ApkSizeTask extends DefaultTask {
     @Input File apk
     @OutputFile File outputFile
+    ApkSizeExtension extension
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy:14]: nit: newline between

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy (hunk 2 above)
Comment (vanniktech) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy:39]: 👍

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy [hunk 7]
```diff
@@ -33,9 +35,20 @@ class ApkSizeTask extends DefaultTask {
...
+        failBuildMaxMethods(apkSize, extension.maxApkSize)
     }
 
-    def withStyledOutput(@ClosureParams(value = SimpleType, options = ['org.gradle.api.logging.Logger']) Closure closure) {
+    def withStyledOutput(@ClosureParams(value = SimpleType, options = ["org.gradle.api.logging.Logger"]) Closure closure) {
         closure(getLogger())
     }
+
+    /**
+     * Fails the build when a user specifies a "max apk size" for their current build.
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy]: not needed

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy [hunk 8]
```diff
@@ -33,9 +35,20 @@ class ApkSizeTask extends DefaultTask {
...
+     */
+    def failBuildMaxMethods(apkSize, maxApkSize) {
+        if (extension.maxApkSize > 0 && apkSize > extension.maxApkSize) {
+            throw new GradleException(String.format("The current APK is %d, the current max is: %d.", apkSize, maxApkSize))
```
Comment (vanniktech) [inline src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy]: also here add the units

File: src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy (hunk 3 above)
Comment (vanniktech) [reply src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy:24]: It does not throw though, right?

**Review by vanniktech — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy (hunk 7 above)
Comment (jaredsburrows) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy]: You do not want comments?

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy (hunk 8 above)
Comment (jaredsburrows) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy]: kb. ok.

File: src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy (hunk 3 above)
Comment (jaredsburrows) [reply src/test/groovy/com/vanniktech/android/apk/size/ApkSizeTaskSpec.groovy:24]: No but `extension` needs to be initiated.

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy (hunk 6 above)
Comment (jaredsburrows) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy:14]: There weren't new lines between the others.

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeExtension.groovy (hunk 5 above)
Comment (jaredsburrows) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeExtension.groovy]: kb. ok.

File: README.md (hunk 4 above)
Comment (jaredsburrows) [reply README.md]: kb. ok.

File: src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy (hunk 7 above)
Comment (vanniktech) [reply src/main/groovy/com/vanniktech/android/apk/size/ApkSizeTask.groovy]: The method name is enough in this case. No need to add those extra comments.

**Comment by jaredsburrows** [comment]
@vanniktech Any other changes?

**Review by vanniktech — APPROVED** [review]
Cool thanks. Sorry that it took forever 👍

**Comment by jaredsburrows** [comment]
@vanniktech Np. I need to write a few tests for it as well!

**Pull request merged**